[SE-0311 Task Locals] sync functions and async{} inheritance (#1357)
apple/swift-evolution: proposals/0311-task-locals.md

- values stored as task-locals are copied into the task's local storage,
- references stored as task-locals are retained and stored by reference in the task's local storage.

Task local "item" storage allocations are performed using an efficient task local stack-discipline allocator, since it is known that those items can never out-live a task they are set on. This makes slightly cheaper to allocate storage for values allocated this way than going through the global allocator, however task-local storage _should not_ be abused to avoid passing parameters explicitly, because it makes your code harder to reason about due to the "hidden argument" passing rather than plain old parameters in function calls.

> Task-local items which are copied to a different task, i.e. when `async{}` is launches a new un-structured task, have independent lifecycles and attach to the newly spawned task. This means that at the point of creating a new task with `async{}` reference count retains may happen for ref-counted types stored within task-local storage.

Reviewer: ```suggestion
Task-local items which are copied to a different task, i.e. when `async{}` launches a new unstructured task, have independent lifecycles and attach to the newly spawned task. This means that, at the point of creating a new task with `async{}`, reference-counted types stored within task-local storage may be retained.
```

  - tasks must always be able to call `Lib.myValue` and get predictable values back; specifically, this means that a task _must not_ be able to mutate its task-local values — because child tasks run concurrently with it, this would mean that a child task invoking `Lib.myValue` twice, could get conflicting results, leading to a confusing programming model
  - **conclusion**: task-local storage must be initialized at task creation time and cannot be mutated, values may only be "bound" by creating new scopes/tasks.

> Note: This approach is similar to how Go's `Context` objects work -- they also cannot be mutated, but only `With(...)` copied, however the copies actually form a chain of contexts, all pointing to their parent context. In Swift, we simply reuse the Concurrency model's inherent `Task` abstraction to implement this pattern.

> #### Task-locals in contexts where no Task is available

Reviewer: Table of contents:

```markdown
    - [Task-locals in contexts where no Task is available](#task-locals-in-contexts-where-no-task-is-available)
```

> Task-locals also are able to function in contexts where no Task is available, they function just as a "dynamic scope" and simply utilize a single thread-local variable to store the task-local storage, rather than forming the chain of storages as it is the case when tasks are available.

Reviewer: ```suggestion
Task-locals are also able to function in contexts where no Task is available, they function just as a "dynamic scope" and simply utilize a single thread-local variable to store the task-local storage, rather than forming the chain of storages as is the case when tasks are available.
```

> The only context in which a Task is not available to the implementation are synchronous functions that were called from the outside of swift concurrency. These functions are very rare but can happen, e.g. when a callback is invoked by a C library on some thread that it managed itself. 

Reviewer: ```suggestion
The only context in which a Task is not available to the implementation are synchronous functions that were called from the outside of Swift Concurrency. These functions are very rare but can happen, e.g. when a callback is invoked by a C library on some thread that it managed itself. 
```

The following API continues to work as expected in those situations:

```swift
func synchronous() {
  withUnsafeCurrentTask { task in 
    assert(task == nil) // no task is available!
  }
  
  Example.$local.withValue(13) { 
    other()
  }
}

func other() {
  print(Example.local) // 13, works as expected
}
```

> The only Swift Concurrency API that is possible to be called from such synchronous function and _will_ inherit task local values is `async{}`, and it will copy any task local values encounterted as usual. This makes for a good inter-op story even with legacy libraries -- we never have to worry if we are on a thread owned by the Swift Concurrency runtime or not, things continue to work as expected.

Reviewer: ```suggestion
The only Swift Concurrency API that can be called from such synchronous functions and _will_ inherit task-local values is `async{}`, and it will copy any task-local values encountered as usual. This makes for a good interoperability story even with legacy libraries -- we never have to worry if we are on a thread owned by the Swift Concurrency runtime or not, and things continue to work as expected.
```
